smart-freight-center/sfc-backend
# feat: allows users to re-create shipment after they are revoked

## Description
This PR makes it possible to reshare shipment footprints after you revoke access to their access.

### How to test it

- Share a footprint 
- Try to share a footprint with the same shipment id(You should get an error and 409)
- Revoke access to that shipment(This should return a 200)
- Try to revoke access to the same shipment(You should get a 404)
- Share a footprint with the same shipmentId(It should succeed)
- When you call get shared emissions, it should return a list of emissions that only a contract definition(I added a `sharedWith` field just the spice things up)

## Approach
It does this by appending a `timestamp` to each asset that is created. This timestamp shows us the exact point that the assets were shared. This makes it possible for us to shared assets with the same shipment

When sharing an asset, however, we now have to check if an asset with that `{shipment}:{timestamp}` has already been (that is if the contract definition exists). If so we throw a `409` with msg `shipment with that id already exists`. 

When the user revokes access to a shipment, we now delete all the contract definitions of assets whose id starts with `{shipment}`. This enables us to revoke access to that data. 

Finally, I made modifications to the `Get shared emissions` endpoint such that it returns only shipments that have not yet been revoked.

### Open Questions and Pre-Merge TODOs
One quick thing to note here is that I added an env variable, `SUPPORTED_EDC_FILTER_OPERATORS`, that would help us know if a particular filter is supported or not. If it is not supported then we simply ignore that filter.

This means that if our connectors support the `LIKE` operator we can use them, if not we perform the filtering with NodeJS
